### PR TITLE
Fix bug: DLL address still use ImageBase address in spite of changing lib_base.

### PR DESCRIPTION
**Pull request by nacayoshi00**
…  lib_base

### Which kind of PR do you create?

- [ x ] This PR only contains minor fixes.
- [ ] This PR contains major feature update.
- [ ] This PR introduces a new function/api for Qiling Framework.

### Coding convention?

- [ x ] The new code conforms to Qiling Framework naming convention.
- [ ] The imports are arranged properly.
- [ ] Essential comments are added.
- [ ] The reference of the new code is pointed out.

### Extra tests?

- [ x ] No extra tests are needed for this PR.
- [ ] I have added enough tests for this PR.
- [ ] Tests will be added after some discussion and review.

### Changelog?

- [ ] This PR doesn't need to update Changelog.
- [ x ] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [ x ] The target branch is dev branch.

### One last thing

- [ x ] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

### CODE REVIEW

**Comment by xwings** [comment]
There are some test with fixed based DLL address. Will you be able to update the test ?

**Comment by nacayoshi00** [comment]
I understood. I will add the test to test_pe.py. Please be patient.

**Comment by nacayoshi00** [comment]
I updated test_pe.py to check the fix of lib address relocation. 
And I also update 2 fixes as below:
 - Implement PE loader to load all of the DLLs recursively when PE loaded.
 - Implement PE loader to be able to load API DLL (like api-ms-xxxx.dll).

I added a sample for the test. So Should I send a pull request to https://github.com/qilingframework/rootfs/tree/master?

**Comment by xwings** [comment]
> I updated test_pe.py to check the fix of lib address relocation.
> And I also update 2 fixes as below:
> 
> * Implement PE loader to load all of the DLLs recursively when PE loaded.
> * Implement PE loader to be able to load API DLL (like api-ms-xxxx.dll).
> 
> I added a sample PE file for the test. So Should I send a pull request to https://github.com/qilingframework/rootfs/tree/master?

Yes. If there is a new exe, you need to update rootfs

**Comment by nacayoshi00** [comment]
OK, I sent the pull request of rootfs. And fix some code to pass the test

File: examples/scripts/dllscollector.bat [hunk 1]
```diff
@@ -55,15 +55,7 @@ if exist %WINDIR%\SysWOW64\winhttp.dll xcopy /f /y %WINDIR%\SysWOW64\winhttp.dll
...
-if exist %WINDIR%\SysWOW64\downlevel\api-ms-win-crt-heap-l1-1-0.dll xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-ms-win-crt-heap-l1-1-0.dll "examples\rootfs\x86_windows\Windows\System32\"
-if exist %WINDIR%\SysWOW64\downlevel\api-ms-win-core-synch-l1-2-0.dll xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-ms-win-core-synch-l1-2-0.dll "examples\rootfs\x86_windows\Windows\System32\"
-if exist %WINDIR%\SysWOW64\downlevel\api-ms-win-core-fibers-l1-1-1.dll xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-ms-win-core-fibers-l1-1-1.dll "examples\rootfs\x86_windows\Windows\System32\"
-if exist %WINDIR%\SysWOW64\downlevel\api-ms-win-core-localization-l1-2-1.dll xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-ms-win-core-localization-l1-2-1.dll "examples\rootfs\x86_windows\Windows\System32\"
-if exist %WINDIR%\SysWOW64\downlevel\api-ms-win-core-sysinfo-l1-2-1.dll xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-ms-win-core-sysinfo-l1-2-1.dll "examples\rootfs\x86_windows\Windows\System32\"
+xcopy /f /y %WINDIR%\SysWOW64\downlevel\api-*.dll "examples\rootfs\x86_windows\Windows\System32\"
```
Comment (xwings) [inline examples/scripts/dllscollector.bat:58]: I think we need to limit and copy only needed DLL. It seems the test took 4 hours due to additional dlls

**Comment by nacayoshi00** [comment]
OK, I updated examples/scripts/dllscollector.bat to copy only needed DLL.

**Comment by xwings** [comment]
Redoing the test and see how long will it take to do the test.

File: qiling/loader/pe.py [hunk 2]
```diff
@@ -575,28 +580,98 @@ def load(self):
...
-                            try:
-                                addr = self.import_address_table[dll_name][imp.name]
-                            except KeyError:
-                                self.ql.log.debug("Error in loading function %s" % imp.name.decode())
-                        else:
-                            addr = self.import_address_table[dll_name][imp.ordinal]
-
-                        if self.ql.archtype == QL_ARCH.X86:
-                            address = self.ql.pack32(addr)
+            # load all DLL files linked from the target PE file recursively and make IAT of all dlls.
```
Comment (xwings) [inline qiling/loader/pe.py:585]: Here onwards

The test before this take around 15-20 min. With the new implementation. It takes around 4 hours to complete the test.

We need to find a better way, This is way too slow.

One method is to cache like how we cache the dll address. maybe we can do the same

File: qiling/loader/pe.py [hunk 3]
```diff
@@ -96,6 +96,11 @@ def load_dll(self, dll_name, driver=False):
                 self.ql.log.warning(f'Warnings while loading {path}:')
                 for warning in warnings:
                     self.ql.log.warning(f' - {warning}')
+
+            # [Room for Improvement] too much time when kernelbase.dll is loaded. 
+            self.ql.log.debug('relocate {}, {:x}'.format(dll_name, self.dll_last_address))
```
Comment (xwings) [inline qiling/loader/pe.py:101]: We need to standardize all the quote. Double quote. " and not '

**Comment by nacayoshi00** [comment]
I see, but the result was unexpected for me because the test took only 1638s on my laptop PC. 
I think the reason why the test is so long is below:
 - All DLLs are loaded at first even if some DLLs may not be used. 
 - DLL reload_image() may take quite long
 
Since I think these fixes are needed because emulation fails when DLL's address is not resolved, I have another solution to solve the problem.
1. Make resolve_dll_address function that resolves address when emulator encounters unmapped address.
2. Register the function to hook_address.

If u r OK, I'll try to update.

**Comment by xwings** [comment]
1638s  is one test or all test ?

The test is being done from github CI, so i guess we need to make that as benchmark.

I guess your idea is fine. We need to fix CI result. Else gonna be bad if we need to spend 4 hours for each test. Since currently only 10min for all test.

**Comment by nacayoshi00** [comment]
It took 1638s  to finish all the tests.
OK, I'll try to fix it. Please be patient.

**Comment by xwings** [comment]
Sure. Will be waiting :)

**Comment by nacayoshi00** [comment]
I'm sorry for late the response. I update these files to load dll dynamically. 
But the problem still remained because the reason why take quite so long is to relocate some dlls such as shell32.dll (see below log.). So I add libcache=True in all_tests of test_pe.py.

`>>> timeit.timeit("import pefile;pefile.PE('qiling/examples/rootfs/x86_windows/Windows/System32/shell32.dll').relocate_image(0x7fff00000)", number=1)
205.60229980000076`

I checked the time dramatically reduced in my PC when libcache is enabled. 
 - previous ver:1638s
 - new version(libcache off): 1000s
 - new version(libcache on): 80s)

**Comment by nacayoshi00** [comment]
I resolved the conflict and re-uploaded

**Comment by nacayoshi00** [comment]
Could you mind if I ask you to run the workflow? Or should I do something to proceed with the workflow?

**Comment by xwings** [comment]
There is still a conlifct with wscok32

**Comment by nacayoshi00** [comment]
Sorry too late. I removed the conflict. So could you please run the workflow?

**Comment by elicn** [comment]
Hi @nacayoshi00,
Based on the recent changes made to `dev` branch, could you please test to see whether this PR is still relevant?

**Comment by xwings** [comment]
Since author did not repnse since Aug 2021. PR will be close for now.

**Pull request closed without merging**